### PR TITLE
 Create feedback decorator

### PR DESCRIPTION
**Pull request by Twinters007**
The first iteration of a feedback decorator. 

I think paired with an auto-send property that you can use with variables and not methods, networktables will be much smoother

### CODE REVIEW

File: magicbot/magicrobot.py [hunk 1]
```diff
@@ -540,6 +560,11 @@ def _inject(self, n, inject_type, cname, component):
     #        for f in d.keys():
     #            d[f] = f(component)     
 
+    def update_feedback(self):
```
Comment (virtuald) [inline magicbot/magicrobot.py]: Why isn't anything calling this?

File: magicbot/magicrobot.py [hunk 2]
```diff
@@ -17,6 +17,21 @@
 
 __all__ = ['MagicRobot']
 
+def feedback(key=None):
```
Comment (virtuald) [inline magicbot/magicrobot.py]: No docstring? What is this supposed to do? Maybe add an example in the docstring too?

File: magicbot/magicrobot.py (hunk 1 above)
Comment (Twinters007) [inline magicbot/magicrobot.py]: Because I don't know how magicrobot works 😉

File: magicbot/magicrobot.py [hunk 3]
```diff
@@ -17,6 +17,49 @@
...
+    In this example, the NetworkTable key is stored at
+    ``/robot/components/component1/angle``
+    """
+    def decorator(func):
+        nt_key = key
+        if nt_key is None:
+            # If no key is passed, get key from method name.
+            # -1 instead of 1 in case 'get_' is not present,
+            # in which case the key will be the method name
+            nt_key = func.__name__.split('get_')[-1]
```
Comment (auscompgeek) [inline magicbot/magicrobot.py]: This will break if `get_` is a non-prefix substring of the function name.

File: magicbot/magicrobot.py [hunk 4]
```diff
@@ -540,6 +592,11 @@ def _inject(self, n, inject_type, cname, component):
     #        for f in d.keys():
     #            d[f] = f(component)     
 
+    def _update_feedback(self):
+        for (component, cname, name) in self._feedbacks:
+            # Put ntvalue at /robot/components/component/key
+            self.__nt.putValue('/components/{0}/{1}'.format(cname, getattr(component, name).__key__), getattr(component, name)())
```
Comment (auscompgeek) [inline magicbot/magicrobot.py]: Maybe avoid calling `getattr` with the same arguments multiple times. `getattr(...)()` looks particularly ugly.

File: magicbot/magicrobot.py (hunk 3 above)
Comment (Twinters007) [inline magicbot/magicrobot.py]: How so? It should still return whatever comes after 'get_', no?

Comment (auscompgeek) [inline magicbot/magicrobot.py]:
```pycon
>>> 'forget_me'.split('get_')[-1]
'me'
```

Comment (Twinters007) [inline magicbot/magicrobot.py]: In that scenario, I imagine you want `forget_me` to be the method, but what about if the method name was... `reverse_get_speed`. Would the expected key be `speed`? or `reverse_get_speed`

Comment (Twinters007) [inline magicbot/magicrobot.py]: Another problem is what if they use camelCase and not lowercase_underscore?

File: magicbot/magicrobot.py [hunk 5]
```diff
@@ -17,6 +17,49 @@
...
+        if nt_key is None:
+            # If no key is passed, get key from method name.
+            # -1 instead of 1 in case 'get_' is not present,
+            # in which case the key will be the method name
+            nt_key = func.__name__.split('get_')[-1]
+        # Set '__feedback__ attribute to be checked during injection
+        setattr(func, '__feedback__', True)
+        # Store key within the function to avoid using class dictionary
+        setattr(func, '__key__', nt_key)
+        return func
```
Comment (virtuald) [inline magicbot/magicrobot.py]: I feel that this should check the function to see if (a) it's a function and (b) do the function arguments match what is required (eg, only a single argument named 'self').

File: magicbot/magicrobot.py [hunk 6]
```diff
@@ -540,6 +592,12 @@ def _inject(self, n, inject_type, cname, component):
     #        for f in d.keys():
     #            d[f] = f(component)     
 
+    def _update_feedback(self):
+        for (component, cname, name) in self._feedbacks:
+            func = getattr(component, name)
+            # Put ntvalue at /robot/components/component/key
+            self.__nt.putValue('/components/{0}/{1}'.format(cname, func.__key__), func())
```
Comment (virtuald) [inline magicbot/magicrobot.py]: If func throws, then the entire robot will be dead.

Comment (Twinters007) [inline magicbot/magicrobot.py]: That'd be embarrassing

File: magicbot/magicrobot.py [hunk 7]
```diff
@@ -52,7 +52,11 @@ def decorator(func):
             # If no key is passed, get key from method name.
             # -1 instead of 1 in case 'get_' is not present,
             # in which case the key will be the method name
-            nt_key = func.__name__.split('get_')[-1]
+            name = func.__name__
+            if name.startswith('get_'):
+                nt_key = name.split('get_')[1]
```
Comment (auscompgeek) [inline magicbot/magicrobot.py]: This will also still fail if `get_` is a non-prefix substring of `name`.

Comment (Twinters007) [inline magicbot/magicrobot.py]: No, it won't.

Comment (virtuald) [inline magicbot/magicrobot.py]: `nt_key = name[4:]`

Comment (Twinters007) [inline magicbot/magicrobot.py]: 3smart5me

File: magicbot/magicrobot.py (hunk 5 above)
Comment (Twinters007) [inline magicbot/magicrobot.py]: Why does it *need* to be named self?

Comment (virtuald) [inline magicbot/magicrobot.py]: It doesn't *need* to be but there typically aren't reasons for it not to be, so we should enforce it.

Comment (Twinters007) [inline magicbot/magicrobot.py]: Luckily state_machine has some code I can steal

Comment (virtuald) [reply magicbot/magicrobot.py]: Still hasn't been added?

Comment (Twinters007) [reply magicbot/magicrobot.py]: Hm I thought I did...

Comment (Twinters007) [reply magicbot/magicrobot.py]: Lol I did just didn't push

**Pull request merged**